include/nbl/builtin/hlsl/bxdf/cook_torrance_base.hlsl

```diff
@@ -17,95 +17,303 @@ namespace hlsl
 namespace bxdf
 {
 
-// N (NDF), F (fresnel), MT (measure transform, using DualMeasureQuant)
-template<class Config, class N, class F, class MT NBL_PRIMARY_REQUIRES(config_concepts::MicrofacetConfiguration<Config> && ndf::NDF<N> && fresnel::Fresnel<F>)
-struct SCookTorrance
+// N (NDF), F (fresnel)
+template<class Config, class N, class F, bool IsBSDF NBL_STRUCT_CONSTRAINABLE>
+struct SCookTorrance;
+
+template<class Config, class N, class F>
+NBL_PARTIAL_REQ_TOP(config_concepts::MicrofacetConfiguration<Config> && ndf::NDF<N> && fresnel::Fresnel<F>)
+struct SCookTorrance<Config, N, F, false NBL_PARTIAL_REQ_BOT(config_concepts::MicrofacetConfiguration<Config> && ndf::NDF<N> && fresnel::Fresnel<F>) >
 {
     NBL_BXDF_CONFIG_ALIAS(scalar_type, Config);
     NBL_BXDF_CONFIG_ALIAS(vector2_type, Config);
     NBL_BXDF_CONFIG_ALIAS(vector3_type, Config);
+    NBL_BXDF_CONFIG_ALIAS(ray_dir_info_type, Config);
     NBL_BXDF_CONFIG_ALIAS(isotropic_interaction_type, Config);
     NBL_BXDF_CONFIG_ALIAS(anisotropic_interaction_type, Config);
     NBL_BXDF_CONFIG_ALIAS(sample_type, Config);
+    NBL_BXDF_CONFIG_ALIAS(spectral_type, Config);
+    NBL_BXDF_CONFIG_ALIAS(quotient_pdf_type, Config);
```

Comment:
the stuff that doesn't metnion the cache could be one big macro in BxDF Config header